libass/ass_render.c

```diff
@@ -2462,14 +2469,17 @@ size_t ass_composite_construct(void *key, void *value, void *priv)
         } else
             copy_bitmap(render_priv->engine, &v->bm_s, &v->bm);
 
+        if (!(flags & FILTER_FILL_IN_SHADOW) && !no_blur)
+            fix_outline(&v->bm, &v->bm_o);
```

Comment:
~~Should this say `bm_s` instead of `bm_o`?~~

Scratch that. Why is this block here at all? If `!no_blur`, nothing should be done.

Or, well, you could leave this here (with `bm_s`) and remove the “In this case no shadow is supposed to be rendered.” check on line 2246. That check happens to save us some work copying & subtracting bitmaps, so we can keep it, or—better yet—we can remove that check, remove this block, and add `if (flags & FILTER_FILL_IN_SHADOW)` to the `else` a couple lines above here. `v` is zero-initialized, so if the flags specify there’s no border width, no 0-width style 3 border, and no filled-in shadow, `v->bm_s.buffer` will remain `NULL` and there will be no shadow bitmap, just as desired.